resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -829,50 +968,175 @@ void add(DnsRecord r) {
+            InetAddress resolved = decodeAddress(r, nsName, parent.isDecodeIdn());
+            if (resolved == null) {
+                // Could not parse the address, just ignore.
+                return;
+            }
+
             while (serverName != null) {
-                if (!serverName.removed && serverName.nsName.equalsIgnoreCase(nsName)) {
-                    serverName.removed = true;
-                    return serverName;
+                if (serverName.nsName.equalsIgnoreCase(nsName)) {
+                    if (serverName.address != null) {
+                        // We received multiple ADDITIONAL records for the same name.
+                        // Search for the last we insert before and then append a new one.
+                        while (serverName.next != null && serverName.next.isCopy) {
+                            serverName = serverName.next;
+                        }
+                        AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
+                        server.next = serverName.next;
+                        serverName.next = server;
+                        serverName = server;
+
+                        servers++;
+                    }
+                    // We should replace the TTL if needed with the one of the ADDITIONAL record so we use
+                    // the smallest for caching.
+                    serverName.update(parent.newRedirectServerAddress(resolved), r.timeToLive());
+                    return;
                 }
                 serverName = serverName.next;
             }
-            return null;
         }
 
-        int size() {
-            return count;
+        // Now handle all AuthoritativeNameServer for which we had no ADDITIONAL record
+        void handleWithoutAdditionals(DnsNameResolver parent, DnsCache cache) {
+            AuthoritativeNameServer serverName = head;
+
+            while (serverName != null) {
+                if (serverName.address == null) {
+                    // Try to resolve via cache
+
+                    List<? extends DnsCacheEntry> entries = cache.get(serverName.nsName, null);
+                    if (entries == null || entries.isEmpty()) {
+                        // These will be resolved on the fly if needed.
+                        serverName.address = InetSocketAddress.createUnresolved(
+                                serverName.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                        servers++;
+                    } else {
+                        // We have some entries in the cache.
+                        for (int i = 0; i < entries.size(); i++) {
+                            InetAddress address = entries.get(i).address();
+                            if (address != null) {
+                                if (serverName.address != null) {
+                                    AuthoritativeNameServer server = new AuthoritativeNameServer(serverName);
```

Comment:
IIUC with the "unrolled loop" this condition `serverName.address != null` can no longer occur? if so this code block can be removed.